File: src/main/java/net/poweredbyscience/truantilag/TruAntiLag.java
package net.poweredbyscience.truantilag;

import org.bukkit.Bukkit;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.PluginDescriptionFile;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitRunnable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.logging.Level;

/**
 * Created by Lax on 1/29/2017.
 */
public class TruAntiLag extends JavaPlugin {

    private ArrayList<String> a = new ArrayList<String>();

    public void onEnable() {
        loadBlacklist();
        getLogger().log(Level.WARNING, "Cracked by LaxWasHere@SpigotMC");
        new BukkitRunnable() {
            @Override
            public void run() {
                getLogger().log(Level.WARNING, "Disabling Lag");
                disableLag();
            }
        }.runTaskLater(this, 600); //Make sure all plugins are loaded
    }

    public void loadBlacklist() {
        for (String s : getBlacklistconfig().getConfigurationSection("name").getKeys(false)) {
            a.add(s.toLowerCase());
        }
        for (String s : getBlacklistconfig().getConfigurationSection("authors").getKeys(false)) {
            a.add(s.toLowerCase());
        }
        for (String s : getBlacklistconfig().getConfigurationSection("mains").getKeys(false)) {
            a.add(s.replace("+",".").toLowerCase());
        }
    }

    public void disableLag() {
        for (Plugin p : Bukkit.getPluginManager().getPlugins()) {
            PluginDescriptionFile pds = p.getDescription();

            if (a.contains(pds.getName().toLowerCase()) && p.isEnabled()) {
                Bukkit.getPluginManager().disablePlugin(p);
            }

            if(a.contains(pds.getMain()) && p.isEnabled()) {
                Bukkit.getPluginManager().disablePlugin(p);
            }

            for (String s : pds.getAuthors()) {
                if (a.contains(s.toLowerCase()) && p.isEnabled()) {
                    Bukkit.getPluginManager().disablePlugin(p);
                }
            }
        }
    }

    public FileConfiguration getBlacklistconfig() {
        try {
            FileConfiguration blacklist = new YamlConfiguration();
            blacklist.load(getClassLoader().getResourceAsStream("laggers.yml"));
            return blacklist;
        } catch (IOException e) {
            e.printStackTrace();
        } catch (InvalidConfigurationException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: src/main/java/net/poweredbyhate/truantilag/TruAntiLag.kt
package net.poweredbyhate.truantilag

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import khttp.get
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitRunnable
import org.json.JSONObject
import java.net.ProxySelector
import java.util.*
import java.util.logging.Level

/**
 * Created by Lax on 1/29/2017.
 * Fixed by Redrield on 2017-10-09
 */
class TruAntiLag : JavaPlugin() {

    private val a = ArrayList<String>()

    companion object {
        lateinit var instance: TruAntiLag
    }

    private val blacklistConfig by lazy {
        get("https://raw.githubusercontent.com/LaxWasHere/TruAntiLag/master/laggers.json").jsonObject.toGson()
    }

    override fun onEnable() {
        instance = this;
        loadBlacklist()
        logger.log(Level.WARNING, "Cracked by LaxWasHere@SpigotMC")
        logger.log(Level.WARNING, "Mega cracked by Redrield@SpigotMC")
        object : BukkitRunnable() {
            override fun run() {
                logger.log(Level.WARNING, "Disabling Lag")
                disableLag()
            }
        }.runTaskLater(this, 600) //Make sure all plugins are loaded
        ProxySelector.setDefault(BadDetector(ProxySelector.getDefault()))
        if (!Bukkit.getVersion().contains("git-Paper")) {
            logger.log(Level.WARNING, "For a TruAntiLag xPerience , use Paper (https://paper.emc.gs)")
        }
    }


    private fun loadBlacklist() {
        blacklistConfig["name"].asJsonObject.entrySet().mapTo(a) { (key, _) -> key.toLowerCase() }
        blacklistConfig["authors"].asJsonObject.entrySet().mapTo(a) { (key, _) -> key.toLowerCase() }
        blacklistConfig["mains"].asJsonObject.entrySet().mapTo(a) { (key, _) -> key.toLowerCase() }
    }

    private fun disableLag() {
        for (p in Bukkit.getPluginManager().plugins) {
            val pds = p.description

            if (a.contains(pds.name.toLowerCase()) && p.isEnabled) {
                Bukkit.getPluginManager().disablePlugin(p)
            }

            if (a.contains(pds.main) && p.isEnabled) {
                Bukkit.getPluginManager().disablePlugin(p)
            }

            pds.authors
                    .filter { a.contains(it.toLowerCase()) && p.isEnabled }
                    .forEach { Bukkit.getPluginManager().disablePlugin(p) }
        }
    }
}

fun JSONObject.toGson(): JsonObject {
    return JsonParser().parse(toString()).asJsonObject

}
